server/app/src/main/java/com/google/heartrate/wearos/app/MainActivity.java

```diff
@@ -18,66 +25,76 @@
+        @Override
+        public void onServiceConnected(ComponentName className, IBinder service) {
+            BluetoothService.BluetoothServerBinder serverBinder = (BluetoothService.BluetoothServerBinder) service;
+            serverBinder.getService().registerGattServiceHandler(heartRateServiceRequestHandler);
+        }
+
+        @Override
+        public void onServiceDisconnected(ComponentName className) { }
+    };
+
     @Override
     protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         setContentView(R.layout.activity_main);
-
-        /* Not to fall into doze and suspend mode */
-        setAmbientEnabled();
-
         mTextView = findViewById(R.id.text);
 
-        startHeartRateServer();
-    }
+        PowerManager powerManager = (PowerManager) getSystemService(Context.POWER_SERVICE);
+        wakeLock = powerManager.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, TAG);
+        wakeLock.acquire();
 
-    /**
-     * Start server with heart rate service hosted.
-     */
-    private void startHeartRateServer() {
+        heartRateSensorListener = new HeartRateSensorListener(this);
         try {
-            heartRateSensorListener = new HeartRateSensorListener(this);
-            heartRateSensorListener.registerSubscriber(this);
-
-            bluetoothServer = new BluetoothServer(this);
+            heartRateSensorListener.startMeasure();
+            heartRateServiceRequestHandler = new HeartRateServiceRequestHandler(heartRateSensorListener);
 
-            HeartRateServiceRequestHandler heartRateServiceRequestHandler
-                    = new HeartRateServiceRequestHandler(heartRateSensorListener);
-            bluetoothServer.registerGattServiceHandler(heartRateServiceRequestHandler);
-
-            bluetoothServer.start();
-        } catch (GattException e) {
+            startForegroundService(new Intent(this, BluetoothService.class));
+        } catch (SensorException e) {
             e.printStackTrace();
         }
     }
 
     @Override
-    protected void onResume() {
-        bluetoothServer.registerReceiver();
-        super.onResume();
+    protected void onStart() {
+        super.onStart();
+
+        heartRateSensorListener.registerSubscriber(this);
+        bindService(new Intent(this, BluetoothService.class), connection, Context.BIND_AUTO_CREATE);
     }
 
     @Override
-    protected void onPause() {
-        bluetoothServer.unregisterReceiver();
-        super.onPause();
+    protected void onStop() {
+        super.onStop();
+
+        unbindService(connection);
+        heartRateSensorListener.unregisterSubscriber(this);
     }
 
     @Override
     protected void onDestroy() {
-        bluetoothServer.stop();
-        heartRateSensorListener.unregisterSubscriber(this);
         super.onDestroy();
+        stopService(new Intent(this, BluetoothService.class));
+        heartRateSensorListener.stopMeasure();
+
+        wakeLock.release();
```

Comment:
Are you sure you want to hold the wakelock in the activity rather than the service?

Reply:
You are right, moved it to service